app/migrations/Version20170401000000.php

```diff
@@ -0,0 +1,99 @@
+<?php
+
+/**
+ * @copyright   2017 Mautic Contributors. All rights reserved
+ * @author      Mautic
+ *
+ * @link        http://mautic.org
+ *
+ * @license     GNU/GPLv3 http://www.gnu.org/licenses/gpl-3.0.html
+ */
+
+namespace Mautic\Migrations;
+
+use Doctrine\DBAL\Migrations\SkipMigrationException;
+use Doctrine\DBAL\Schema\Schema;
+use Mautic\CoreBundle\Doctrine\AbstractMauticMigration;
+use Mautic\EmailBundle\Model\EmailModel;
+
+/**
+ * Auto-generated Migration: Please modify to your needs!
+ */
+class Version20170401000000 extends AbstractMauticMigration
+{
+    /**
+     * @param Schema $schema
+     *
+     * @throws SkipMigrationException
+     * @throws \Doctrine\DBAL\Schema\SchemaException
+     */
+    public function preUp(Schema $schema)
+    {
+        $table = $schema->getTable($this->prefix.'email_stats');
+        if ($table->hasColumn('is_clicked') === true) {
+            throw new SkipMigrationException('Schema includes this migration');
+        }
+    }
+
+    /**
+     * @param Schema $schema
+     */
+    public function up(Schema $schema)
+    {
+        $this->addSql("ALTER TABLE {$this->prefix}emails ADD clicked_count INT NOT NULL;");
+        $this->addSql("ALTER TABLE {$this->prefix}email_stats ADD is_clicked TINYINT(1) NOT NULL, ADD date_clicked DATETIME DEFAULT NULL COMMENT '(DC2Type:datetime)';");
+        $this->addSql("CREATE INDEX stat_email_clicked_search ON {$this->prefix}email_stats (is_clicked);");
+    }
+
+    /**
+     * Update all the channel events with their assigned channels.
+     *
+     * @param Schema $schema
+     */
+    public function postUp(Schema $schema)
+    {
+
+        /** @var EmailModel $emailModel */
+        $emailModel = $this->container->get('mautic.email.model.email');
+        $statRepo   = $emailModel->getStatRepository();
+
+        $logger = $this->container->get('monolog.logger.mautic');
+
+        /** @var \Doctrine\DBAL\Query\QueryBuilder $qb */
+        $qb = $this->connection->createQueryBuilder();
+
+        // get all email_stats where lead and email ids are not null and where there is a page hit
+        $qb->select('es.email_id as emailId, es.id as statId, es.lead_id as leadId')
+            ->addSelect('ph.date_hit as hit')
+            ->from($this->prefix.'email_stats', 'es')
+            ->join('es', 'page_hits', 'ph', 'ph.email_id = es.email_id')
+            ->where('es.email_id IS NOT NULL')
+            ->andWhere('es.lead_id IS NOT NULL')
+            ->andWhere('ph.lead_id = es.lead_id');
+
+        $results = $qb->execute();
+
+        while (($res = $results->fetch()) !== false) {
```

Comment:
I appreciate your optimisation of the query, I really do. The problem is that if a migration takes more than 30 seconds, which is default max_execution_time limit, the migration will fail in the process which causes big troubles for users. This limit applies on update via web UI. CLI has higher limit, but since the web UI migration is there, users will use it. Based on your tests, the limit will be hit with just 200 000 rows in the email_stats table.

Mautic has already bad reputation with upgrade issues and I'm sure this migration will cause a lot of troubles. You may create a pole among the community whether they want stable upgrade or click counts. My guess is that trouble-less upgrade will win. I myself like this PR and want to get it in, but we must do everything to minimise the risk of failed upgrade. So lets optimise it the best we can.

I can see that grouping and counting won't work for this case since you want to upgrade email_stat row with hit date. Take a look at [these code examples](http://docs.doctrine-project.org/projects/doctrine-orm/en/latest/reference/batch-processing.html#id1) from Doctrine's docs and try to apply them here. The important improvement is that the updates will be flushed to database in small batches and not one by one. Also don't forget to detach processed rows to free memory.